pkg/workflow/controllers/parallel_node_reconciler.go

```diff
@@ -0,0 +1,245 @@
+			"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+		)
+		return nil
+	}
+
+	activeChildrenNodes, finishedChildrenNodes, err := it.fetchChildrenNodes(ctx, node)
+	if err != nil {
+		return err
+	}
+	existsChildrenNodes := append(activeChildrenNodes, finishedChildrenNodes...)
+
+	var taskNamesOfNodes []string
+	for _, childNode := range existsChildrenNodes {
+		taskNamesOfNodes = append(taskNamesOfNodes, getTaskNameFromGeneratedName(childNode.GetName()))
+	}
+
+	var tasksToStartup []string
+	if len(relativeComplementSet(taskNamesOfNodes, node.Spec.Tasks)) > 0 {
+		// TODO: check the specific of task and workflow nodes
+		// the definition of Spec.Tasks changed, remove all the existed nodes
+		tasksToStartup = node.Spec.Tasks
+		for _, childNode := range existsChildrenNodes {
+			// best effort deletion
+			err := it.kubeClient.Delete(ctx, &childNode)
+			if err != nil {
+				it.logger.Error(err, "failed to delete outdated child node",
+					"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+					"child node", fmt.Sprintf("%s/%s", childNode.Namespace, childNode.Name),
+				)
+			}
+		}
+	} else {
+		tasksToStartup = relativeComplementSet(node.Spec.Tasks, taskNamesOfNodes)
+	}
+
+	if len(tasksToStartup) == 0 {
+		it.logger.Info("no need to spawn new child node", "node", fmt.Sprintf("%s/%s", node.Namespace, node.Name))
+		return nil
+	}
+
+	parentWorkflow := v1alpha1.Workflow{}
+	err = it.kubeClient.Get(ctx, types.NamespacedName{
+		Namespace: node.Namespace,
+		Name:      node.Spec.WorkflowName,
+	}, &parentWorkflow)
+	if err != nil {
+		it.logger.Error(err, "failed to fetch parent workflow",
+			"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+			"workflow name", node.Spec.WorkflowName)
+		return err
+	}
+
+	childrenNodes, err := renderNodesByTemplates(&parentWorkflow, &node, tasksToStartup...)
+	if err != nil {
+		it.logger.Error(err, "failed to render children childrenNodes",
+			"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name))
+		return err
+	}
+
+	// TODO: emit event
+	var childrenNames []string
+	for _, childNode := range childrenNodes {
+		err := it.kubeClient.Create(ctx, childNode)
+		if err != nil {
+			it.logger.Error(err, "failed to create child node",
+				"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+				"child node", childNode)
+			return err
+		}
+		childrenNames = append(childrenNames, childNode.Name)
+	}
+	it.logger.Info("parallel node spawn new child node",
+		"node", fmt.Sprintf("%s/%s", node.Namespace, node.Name),
+		"child node", childrenNames)
+
+	return nil
+}
+
+func getTaskNameFromGeneratedName(generatedNodeName string) string {
+	index := strings.LastIndex(generatedNodeName, "-")
+	if index < 0 {
+		return generatedNodeName
+	}
+	return generatedNodeName[:index]
+}
+
+// relativeComplementSet return the set of elements which contained in former but not in latter
+func relativeComplementSet(former []string, latter []string) []string {
```

Comment:
That is call `set difference` not `complement`, if I haven't got it wrong.

Reply:
I want to present that: https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement

Reply:
Oh, mb.

Reply:
It seems relativeComplementSet indicate B-A while set difference is A-B, I thought you mean A-B?

Reply:
Yeah, you are right!

Reply:
fixed, PTAL